pkg/sql/virtual_table.go

```diff
@@ -24,20 +25,79 @@ import (
 // nil). If there is an error, then return (nil, error).
 type virtualTableGenerator func() (tree.Datums, error)
 
+type virtualTableGeneratorResponse struct {
+	datums tree.Datums
+	err    error
+}
+
+var errRowsDone = errors.New("predefined error that the generator has been canceled")
+
+// setupGenerator takes in a worker that generates rows eagerly and transforms
+// it into a lazy row generator. It returns three functions:
+// * cleanup: Performs all cleanup. May only be called once.
+// * startGenerator: The generator version of the input worker. Intended to
+//   be started as a goroutine.
+// * next: A handle that can be called to generate a row from the worker. Next
+//   cannot be called once cleanup has been called.
+func setupGenerator(
+	worker func(addRow func(...tree.Datum) error) error,
+) (cleanup func(), startGenerator func(), next virtualTableGenerator) {
+	done, send := make(chan struct{}), make(chan virtualTableGeneratorResponse)
+	cleanup = func() {
+		close(done)
+	}
+	addRow := func(datums ...tree.Datum) error {
+		// We need another select here so that in case cleanup has
```

Comment:
hmm, this  first select doesn't seem necessary. can you describe the race a bit more closely?

Reply:
Yeah i don't think it is necessary, it was to catch people calling next after cleanup which is illegal anyway.